src/openmrs-fetch.ts

```diff
@@ -0,0 +1,145 @@
+    fetchInit.headers.Accept = "application/json";
+  }
+
+  if (fetchInit.headers.Accept === null) {
+    delete fetchInit.headers.Accept;
+  }
+
+  /* We capture the stacktrace before making the request, so that if an error occurs we can
+   * log a full stacktrace that includes the code that made the request and handled the response
+   * Otherwise, we could run into situations where the stacktrace doesn't even show which code
+   * called @openmrs/api.
+   */
+  const requestStacktrace = Error();
+
+  return window.fetch(url, fetchInit as RequestInit).then(response => {
+    if (response.ok) {
+      if (response.status === 204) {
+        /* HTTP 204 - No Content
+         * We should not try to download the empty response as json. Instead,
+         * we return null since there is no response body.
+         */
+        return null;
+      } else {
+        // HTTP 200s - The request succeeded
+        return response.json();
+      }
+    } else {
+      /* HTTP response status is not in 200s. Usually this will mean
+       * either HTTP 400s (bad request from browser) or HTTP 500s (server error)
+       * Our goal is to come up with best possible stacktrace and error message
+       * to help developers understand the problem and debug
+       */
+
+      // Attempt to download a response body, if it has one
+      return response.text().then(
+        responseText => {
+          let responseBody = responseText;
+          try {
+            responseBody = JSON.parse(responseText);
+          } catch (err) {
+            // Server didn't respond with json, so just go with the response text string
+          }
+
+          /* Make the fetch promise go into "rejected" status, with the best
+           * possible stacktrace and error message.
+           */
+          throw new OpenmrsFetchError(
+            url,
+            response,
+            responseBody,
+            requestStacktrace
+          );
+        },
+        err => {
+          /* We weren't able to download a response body for this error.
+           * Time to just give the best possible stacktrace and error message.
+           */
+          throw new OpenmrsFetchError(url, response, null, requestStacktrace);
+        }
+      );
+    }
+  });
+}
+
+class OpenmrsFetchError extends Error {
+  constructor(
+    url: string,
+    response: Response,
+    responseBody: ResponseBody | null,
+    requestStacktrace: Error
+  ) {
+    super();
+    this.message = `Server responded with ${response.status} (${response.statusText}) for url ${url}. Check err.responseBody or network tab in dev tools for more info`;
+    requestStacktrace.message = this.message;
+    this.responseBody = responseBody;
+    this.stack = `Stacktrace for outgoing request:\n${requestStacktrace.stack}\nStacktrace for incoming response:\n${this.stack}`;
+  }
+  responseBody: string | FetchResponseJson | null;
+}
+
+type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
```

Comment:
Look like Omit is included in TS 3.5 https://devblogs.microsoft.com/typescript/announcing-typescript-3-5/

Reply:
I'll look into this in a bit

Reply:
Yeah my IDE was freaking out but typescript@3.5 is not. So we can use the built-in Omit. I just updated this. Good catch, @brandones